src/librustc_driver/lib.rs

```diff
@@ -187,76 +151,291 @@ fn run_compiler(args: &[String]) {
         None => {/* continue */ }
     }
 
-    if sess.unstable_options() {
-        sess.opts.show_span = matches.opt_str("show-span");
-    }
-
-    let r = matches.opt_strs("Z");
-    if r.contains(&("ls".to_string())) {
-        match input {
-            Input::File(ref ifile) => {
-                let mut stdout = old_io::stdout();
-                list_metadata(&sess, &(*ifile), &mut stdout).unwrap();
-            }
-            Input::Str(_) => {
-                early_error("cannot list metadata for stdin");
-            }
-        }
-        return;
-    }
-
     let plugins = sess.opts.debugging_opts.extra_plugins.clone();
-    let control = build_controller(&sess);
+    let control = callbacks.build_controller(&sess);
     driver::compile_input(sess, cfg, &input, &odir, &ofile, Some(plugins), control);
 }
 
-fn build_controller<'a>(sess: &Session) -> CompileController<'a> {
-    let mut control = CompileController::basic();
+// Extract output directory and file from matches.
+fn make_output(matches: &getopts::Matches) -> (Option<Path>, Option<Path>) {
+    let odir = matches.opt_str("out-dir").map(|o| Path::new(o));
+    let ofile = matches.opt_str("o").map(|o| Path::new(o));
+    (odir, ofile)
+}
 
-    if sess.opts.parse_only ||
-       sess.opts.show_span.is_some() ||
-       sess.opts.debugging_opts.ast_json_noexpand {
-        control.after_parse.stop = true;
+// Extract input (string or file and optional path) from matches.
+fn make_input(free_matches: &[String]) -> Option<(Input, Option<Path>)> {
+    if free_matches.len() == 1 {
+        let ifile = &free_matches[0][];
+        if ifile == "-" {
+            let contents = old_io::stdin().read_to_end().unwrap();
+            let src = String::from_utf8(contents).unwrap();
+            Some((Input::Str(src), None))
+        } else {
+            Some((Input::File(Path::new(ifile)), Some(Path::new(ifile))))
+        }
+    } else {
+        None
     }
+}
 
-    if sess.opts.no_analysis || sess.opts.debugging_opts.ast_json {
-        control.after_write_deps.stop = true;
+// A trait for customising the compilation process. Offers a number of hooks for
+// executing custom code or customising input.
+pub trait CompilerCalls<'a> {
+    // Hook for a callback early in the process of handling arguments. This will
+    // be called straight after options have been parsed but before anything
+    // else (e.g., selecting input and output). Return true to terminate compilation,
```

Comment:
I feel like the bool will be confusing (since it could also be "should I continue", rather than "should I stop"), maybe this could be something like `enum Compilation { Continue, Stop }`?